ListNotifier Tests - notificationMode types

KhaderMurtaja opened the pull request:
### Test Results ⬇️ 

<img width="1376" alt="list_notifier_test_results" src="https://github.com/user-attachments/assets/82afa640-5115-4efa-bbf4-a31ac8ece89f">

KhaderMurtaja committed:
pub get

KhaderMurtaja committed:
Correct the comment and mention all notificationMode types

KhaderMurtaja committed:
Write tests for all notificationMode types

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,12 +5,23 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
-/// If you set [notifyIfEqual] to `false` it will compare if a value passed value
-///  passed is equal to the existing value.
-/// like `list[5]=4` if the content at index 4 is equal to 4 and only call
+/// It does not compare the elements on bulk operations
```
Actually this isn't longer the case. for most bulk operations it will compare elements.

escamoteur replied:
The only exceptions are 
![grafik](https://github.com/escamoteur/listenable_collections/assets/16635729/86c530a3-a7bb-4c47-9acc-39e16b6cae11)
perhaps we should list this functionnames there.  Most probably these functions will always change the List. 
I could not move the comments before the method name because then the inherited API doc of List won't appear when you hover over a method

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,12 +5,23 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
+///
+/// If you set [notificationMode] to [normal] `ListNotifier` will compare if a
+/// value passed is equal to the existing value.
+/// Like `list[5]=4` if the content at index 4 is equal to 4 and only call
```
Like in `list[5]=4` it tests if the content at index 4 is equal to 4 and only call

escamoteur reviewed (changes requested):
Thanks for your work, just some minor changes, then we are done

KhaderMurtaja committed:
Use single quotes instead of double

KhaderMurtaja committed:
Use explicit types

KhaderMurtaja committed:
Remove the unnecessary use of parentheses.

KhaderMurtaja committed:
Fix typo

KhaderMurtaja committed:
Use explicit type for the result list

KhaderMurtaja committed:
Use explicit type arguments

KhaderMurtaja committed:
The value of the notificationMode is redundant because it matches the default value.

KhaderMurtaja committed:
Add comments regarding the functions and the class in general

KhaderMurtaja committed:
Use explicit type argument(s) for 'IterableEquality'

KhaderMurtaja committed:
Format code

KhaderMurtaja committed:
Add @override annotation as it overrides an inherited method

KhaderMurtaja committed:
Fix typo

KhaderMurtaja committed:
Removed the Unnecessary duplication of receiver, and use cascade notation

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
-/// value passed is equal to the existing value.
-/// Like `list[5]=4` if the content at index 4 is equal to 4 and only call
-/// `notifyListeners` if they are not equal.
+/// If you want to compare the elements you can use the
```
why shortening this? The original comment might be improved but explained it better

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
-/// If you set [notificationMode] to [manual] you have to call `notifyListeners`
-/// manually to notify all listeners.
-/// Like `list.notifyListeners();` if you want to notify all listeners.
+/// If you want to prevent the list from notifying listeners you can use the
```
same, improve the original but don't reduce information

KhaderMurtaja replied:
have done that.

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
-/// If you set [notificationMode] to [always] `ListNotifier` will always notify
-/// all listeners if the list changes.
-/// Like `list[5]=4` will always notify all listeners.
+/// If you want to notify listeners on every change you can use the
```
same

escamoteur replied:
include the example again

KhaderMurtaja replied:
@escamoteur, it's already there at lines `[19, 22]` 🙋🏼‍♂️ 

![screenshot](https://github.com/escamoteur/listenable_collections/assets/29831987/aac2042f-fe85-4749-9fad-8627b646dcd4)

escamoteur replied:
It's not bad if we repeat some information

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
+/// If you want to notify listeners on every change you can use the
+/// `CustomNotifierMode.always` mode.
 ///
-/// To allow atomic changes `ListNotifier` supports a single level of
```
why remove this?

KhaderMurtaja replied:
It was mistakenly, sorry!

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -5,23 +5,24 @@ import 'package:flutter/foundation.dart' show ChangeNotifier, ValueListenable;
 ///
-/// To allow atomic changes `ListNotifier` supports a single level of
-/// transactions
+/// The functions that will always notify listeners unless [notificationMode]
```
I would add `will always notifiy listeners without comparing the existing value`

KhaderMurtaja replied:
Done ✅

escamoteur replied:
" without comparing the existing value" still  missing

KhaderMurtaja replied:
@escamoteur, it's already there at lines `[35, 35]` 🙋🏼‍♂️ 

![screenshot](https://github.com/escamoteur/listenable_collections/assets/29831987/43eb5185-d50e-4c76-8bf0-dfd0faa65dd1)

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -32,10 +33,10 @@ class ListNotifier<T> extends DelegatingList<T>
-  /// the default is [always].
-  /// [customEquality] can be used to set your own criteria for comparing when choosing
-  /// [normal] as [notifierMode].
+  /// why their UI doesn't update if they haven't overwritten the equality
```
actually the correct way would be `overridden` if we talk about overriding an method.

KhaderMurtaja replied:
Fixed

escamoteur commented on test/list_notifier_test.dart:
```diff
@@ -5,8 +5,8 @@ import '../lib/src/list_notifier.dart';
   group("Tests for the ListNotifier's methods", () {
-    late ListNotifier list;
-    List result = [];
+    late ListNotifier<int> list;
```
yep, good to use types here

escamoteur replied:
As the new default it to notify always we should set mode to normal for the tests.

escamoteur replied:
you didn't address this comment

KhaderMurtaja replied:
@escamoteur, what do you mean by "address this comment"? 🤔👀

escamoteur replied:
I meant this comment:
>As the new default it to notify always we should set mode to normal for the tests.

if you change that you will have to review the tests again

KhaderMurtaja committed:
Enhance the old original comments

KhaderMurtaja committed:
Fix the comment

escamoteur commented on lib/src/list_notifier.dart:
```diff
@@ -4,13 +4,30 @@ import 'package:collection/collection.dart';
+/// data changes.
+/// It does not compare the elements during bulk operations.
+///
+/// If [notifyIfEqual] is set to `false`, `ListNotifier` will compare the new
```
why do I see again `notifyIfEqual` in the comments, that does no longer exist. We now have the `notificationMode` so correct would be if [notificationMode] is set to [normal].

The default is [always] so that the UI will be updated on every write access of the List

KhaderMurtaja replied:
@escamoteur, fixed it 🙋🏼‍♂️

escamoteur reviewed (changes requested):
still some open points. sorry for nitpicking.
Amal didn't push anything yet.

KhaderMurtaja committed:
Enhance the comments and explain the notificationMode modes

KhaderMurtaja commented:
> still some open points. sorry for nitpicking. Amal didn't push anything yet.

@escamoteur, never mind I enjoy enhancing the documentation 😁, it has to be perfect for better understanding 💯
I have addressed your comments, I hope you find it ready to merge now unless you have more points 🙏🏼

escamoteur reviewed (changes requested):
Docs are fine now, but the tests should use notificationtype.normal which means they have to be tested again then

KhaderMurtaja committed:
Tests for CustomNotifierMode.noraml | notify when value is changed and don't when value doesn't

KhaderMurtaja committed:
Separate notificationMode TEST in separate group

KhaderMurtaja committed:
Write test for customEquality handling null values

KhaderMurtaja commented:
> Docs are fine now, but the tests should use notificationtype.normal which means they have to be tested again then

@escamoteur, All tests are covered now, and even more tests have been added such as: `"customEquality handles null values correctly"`.

Check out the [PR description](https://github.com/escamoteur/listenable_collections/pull/7#issue-2317626275) for the Test Results ✅.

escamoteur reviewed (changes requested):
I fear you misunderstood me.
As most test here were written with the old default of NotificationMode.normal, you should change the Notification mode of the ListNotifier that is used through all of this tests in setup to NotificationMode.Normal
Before you change any of the tests
